## 1. Carga de los datos

Se carga el conjunto de datos Online Retail, que contiene información sobre las transacciones realizadas por una empresa minorista.

El conjunto incluye variables relacionadas con las facturas, productos, cantidades, fechas, precios, clientes y países.

In [1]:
import pandas as pd

data = pd.read_excel("../data/raw/OnlineRetail.xlsx")

### Visualización inicial

Se muestran las primeras filas del conjunto de datos para conocer su estructura y observar el tipo de información disponible en cada registro.

In [2]:
data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


### Dimensiones del conjunto de datos

Se consulta el número de registros y variables que contiene el conjunto de datos original.

In [3]:
data.shape

(541909, 8)

### Información general

Se revisan los nombres de las variables, la cantidad de valores no nulos y los tipos de datos de cada columna.

Esta revisión permite identificar posibles problemas de calidad de datos que deberán considerarse durante la etapa de limpieza.

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


### Identificación de valores faltantes

Se contabilizan los valores faltantes de cada variable para identificar qué columnas requieren atención durante la limpieza de los datos.

In [5]:
data.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

### Valores faltantes en CustomerID

Se contabiliza la cantidad de registros que no contienen un identificador de cliente.

In [6]:
data['CustomerID'].isnull().sum()

135080

### Porcentaje de registros sin CustomerID

Se calcula el porcentaje de registros que no cuentan con un identificador de cliente respecto al total de observaciones.

In [7]:
porcentaje_customerid = data['CustomerID'].isnull().sum() / len(data) * 100
porcentaje_customerid

24.926694334288598

### Análisis de la variable UnitPrice

Se revisan las estadísticas descriptivas de la variable `UnitPrice` para conocer la distribución de los precios por unidad e identificar posibles valores inusuales que puedan afectar el análisis.

In [8]:
data['UnitPrice'].describe()

count    541909.000000
mean          4.611114
std          96.759853
min      -11062.060000
25%           1.250000
50%           2.080000
75%           4.130000
max       38970.000000
Name: UnitPrice, dtype: float64

### Identificación de precios menores o iguales a cero

Se identifican los registros cuyo precio por unidad es menor o igual a cero, con el objetivo de detectar valores inusuales que puedan requerir una revisión antes del análisis.

In [9]:
data[data['UnitPrice'] <= 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,2011-12-08 13:58:00,0.0,NaN,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,2011-12-08 13:58:00,0.0,NaN,United Kingdom
538554,581408,85175,NaN,20,2011-12-08 14:06:00,0.0,NaN,United Kingdom


### Identificación de precios iguales a cero

Se revisan los registros cuyo precio por unidad es exactamente cero para determinar si corresponden a errores, registros especiales o transacciones que puedan requerir un tratamiento específico durante la limpieza.

In [10]:
data[data['UnitPrice'] == 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,2011-12-08 13:58:00,0.0,NaN,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,2011-12-08 13:58:00,0.0,NaN,United Kingdom
538554,581408,85175,NaN,20,2011-12-08 14:06:00,0.0,NaN,United Kingdom


### Revisión de precios negativos

Se revisan específicamente los registros con precios por unidad negativos para determinar su naturaleza y evaluar si deben considerarse en el análisis de ventas.

In [11]:
data[data['UnitPrice']<0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


### Interpretación de los precios negativos

La revisión de los registros con precios negativos mostró que corresponden a movimientos identificados como `Adjust bad debt`, por lo que no representan ventas ordinarias de productos.

Debido a esta característica, estos registros se consideran casos especiales y se excluyen posteriormente del análisis de ventas.

### Valores faltantes en InvoiceNo

In [12]:
data['InvoiceNo'].isnull().sum()

0

### Revisión de precios iguales a cero y descripción faltante

Se revisan los registros con precio por unidad igual a cero y sin descripción para identificar posibles registros especiales o inconsistencias en los datos.

In [13]:
data[(data['UnitPrice'] == 0) & (data['Description'].isnull())]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


### Revisión de registros con información incompleta

Se analizan los registros que presentan simultáneamente un precio por unidad igual a cero y valores faltantes en la descripción y el identificador del cliente.

Esta revisión permite determinar si estos registros contienen información suficiente para ser considerados en el análisis de ventas.

In [14]:
data[(data['UnitPrice'] == 0) & (data['Description'].isnull()) &(data['CustomerID'].isnull())]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535322,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


### Interpretación de los registros con información faltante

La revisión mostró que los registros con `UnitPrice` igual a cero y `Description` faltante también presentan `CustomerID` faltante.

Esto indica que se trata de registros con información incompleta que requieren un tratamiento específico antes de utilizarlos en el análisis de ventas.

### Revisión de registros con precio cero y descripción disponible

Se analizan los registros cuyo precio por unidad es igual a cero pero que cuentan con una descripción.

El objetivo es determinar si estos registros corresponden a productos o transacciones válidas, o si presentan características que justifican su exclusión del análisis.

In [15]:
data[(data['UnitPrice'] == 0) & (~data['Description'].isnull())]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,NaN,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,NaN,United Kingdom
7313,537032,21275,?,-30,2010-12-03 16:50:00,0.0,NaN,United Kingdom
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647.0,Germany
13217,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535336,581213,22576,check,-30,2011-12-07 18:38:00,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,2011-12-08 09:56:00,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,2011-12-08 13:58:00,0.0,NaN,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,2011-12-08 13:58:00,0.0,NaN,United Kingdom


### Identificación de registros con precio cero e información disponible

Se identifican los registros con `UnitPrice` igual a cero que cuentan tanto con una descripción como con un identificador de cliente.

Estos casos se revisan por separado, ya que contienen información suficiente para evaluar si corresponden a transacciones válidas o a registros especiales.

In [16]:
data[(data['UnitPrice'] == 0) & (~data['Description'].isnull()) &(~data['CustomerID'].isnull())].shape[0]

40

### Revisión de registros con precio cero, descripción faltante y cantidad positiva

Se analizan los registros con `UnitPrice` igual a cero, `Description` faltante y una cantidad positiva.

Esta revisión permite identificar si existen movimientos de productos que, a pesar de no tener precio ni descripción, registran unidades asociadas a la transacción.

In [17]:
data[(data['UnitPrice'] == 0) & (data['Description'].isnull()) &(data['Quantity'] > 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
533712,581103,22689,NaN,4,2011-12-07 11:58:00,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


### Definición de referencia para ventas con información incompleta

Se contabilizan los códigos de producto presentes entre los registros con `UnitPrice` igual a cero, `Description` faltante y cantidad positiva.

Esta revisión permite conocer si estos registros se concentran en pocos productos o si corresponden a diferentes códigos de producto.

In [18]:
data[(data['UnitPrice'] == 0) & (data['Description'].isnull()) &(data['Quantity'] > 0)]['StockCode'].value_counts()

StockCode
35965     9
23084     8
22084     7
23348     5
22502     4
         ..
82494L    1
16248B    1
21279     1
21278     1
85175     1
Name: count, Length: 426, dtype: int64

### Definición de referencia para ventas con información completa

Se contabiliza la frecuencia de los códigos de producto en registros que presentan un precio positivo, una descripción disponible y una cantidad positiva.

Este resultado se utiliza como referencia para comparar los registros con precio cero e información incompleta.

In [19]:
data[(data['UnitPrice'] > 0) & (data['Description'].notnull()) & (data['Quantity'] > 0)]['StockCode'].value_counts()

StockCode
85123A      2265
85099B      2112
22423       2017
47566       1706
20725       1595
            ... 
DCGS0004       1
84705C         1
20964          1
72803b         1
23843          1
Name: count, Length: 3922, dtype: int64

### Comparación con productos presentes en ventas normales

Se identifican los registros con precio cero, descripción faltante y cantidad positiva cuyos códigos de producto también aparecen en transacciones consideradas normales.

Esta comparación permite determinar si los registros con información incompleta están asociados a productos que también aparecen en operaciones de venta con información disponible.

In [20]:
data[
    (data['UnitPrice'] == 0) &
    (data['Description'].isnull()) &
    (data['Quantity'] > 0) &
    (data['StockCode'].isin(
        data[
            (data['UnitPrice'] > 0) &
            (data['Description'].notnull()) &
            (data['Quantity'] > 0)
        ]['StockCode']
    ))
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom
4348,536765,84952C,NaN,19,2010-12-02 14:43:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
533712,581103,22689,NaN,4,2011-12-07 11:58:00,0.0,NaN,United Kingdom
535326,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
535332,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
536981,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


### Identificación de códigos de producto repetidos

Se contabiliza la frecuencia de cada `StockCode` dentro de los registros con precio cero, descripción faltante y cantidad positiva que también aparecen en ventas normales.

Este análisis permite identificar qué códigos de producto se repiten con mayor frecuencia dentro de estos registros especiales.

In [21]:
data[
    (data['UnitPrice'] == 0) &
    (data['Description'].isnull()) &
    (data['Quantity'] > 0) &
    (data['StockCode'].isin(
        data[
            (data['UnitPrice'] > 0) &
            (data['Description'].notnull()) &
            (data['Quantity'] > 0)
        ]['StockCode']
    ))
]['StockCode'].value_counts()

StockCode
35965     9
23084     8
22084     7
23348     5
POST      4
         ..
20748     1
20821     1
44265     1
84247L    1
85175     1
Name: count, Length: 410, dtype: int64

### Revisión de los productos asociados

Se revisan algunos de los códigos de producto identificados en los registros con precio cero, descripción faltante y cantidad positiva, comparándolos con las transacciones normales en las que esos mismos códigos presentan una descripción y un precio disponible.

Esta revisión permite comprobar que algunos de estos registros corresponden a productos que también aparecen en operaciones con información completa.

In [22]:
data[
    (data['UnitPrice'] > 0) &
    (data['Description'].notnull()) &
    (data['Quantity'] > 0) &
    (data['StockCode'].isin(
        data[
            (data['UnitPrice'] == 0) &
            (data['Description'].isnull()) &
            (data['Quantity'] > 0)
        ]['StockCode']
    ))
][['StockCode', 'Description', 'UnitPrice']].head(20)

,StockCode,Description,UnitPrice
3,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,3.39
4,84029E,RED WOOLLY HOTTIE WHITE HEART.,3.39
8,22632,HAND WARMER RED POLKA DOT,1.85
19,21777,RECIPE BOX WITH METAL HEART,7.95
28,22726,ALARM CLOCK BAKELIKE GREEN,3.75
31,10002,INFLATABLE POLITICAL GLOBE,0.85
45,POST,POSTAGE,18.00
47,22632,HAND WARMER RED POLKA DOT,1.85
53,37370,RETRO COFFEE MUGS ASSORTED,1.06
55,21071,VINTAGE BILLBOARD DRINK ME MUG,1.06


### Distribución geográfica de los registros identificados

Se analiza el país asociado a los registros con precio cero, descripción faltante y cantidad positiva cuyos códigos también aparecen en ventas normales.

Los resultados muestran que los 576 registros identificados pertenecen a `United Kingdom`.

In [23]:
data[
    (data['UnitPrice'] == 0) &
    (data['Description'].isnull()) &
    (data['Quantity'] > 0) &
    (data['StockCode'].isin(
        data[
            (data['UnitPrice'] > 0) &
            (data['Description'].notnull()) &
            (data['Quantity'] > 0)
        ]['StockCode']
    ))
]['Country'].value_counts()

Country
United Kingdom    576
Name: count, dtype: int64

### Período de tiempo de los registros analizados

Se identifica el período de tiempo cubierto por los registros con precio igual a cero, descripción nula y cantidad positiva.

Esto permite conocer cuándo se presentan estos registros dentro del conjunto de datos y aporta contexto para evaluar su naturaleza.

In [24]:
registros_precio_cero = data[
    (data['UnitPrice'] == 0) &
    (data['Description'].isnull()) &
    (data['Quantity'] > 0)
]

registros_precio_cero['InvoiceDate'].min(), registros_precio_cero['InvoiceDate'].max()

(Timestamp('2010-12-01 11:52:00'), Timestamp('2011-12-08 14:06:00'))

### Revisión de registros con precio cero y descripción faltante

Se identificaron 592 registros con `UnitPrice` igual a cero, `Description` faltante y cantidades positivas.

Se revisó si sus códigos de producto podían asociarse con productos presentes en transacciones normales. De estos registros, 576 correspondieron a códigos que sí aparecían en otras transacciones, mientras que 16 no pudieron asociarse con una descripción válida.

Los 16 registros sin una descripción identificable se consideran casos especiales y se revisarán antes de definir su tratamiento durante la limpieza.

In [25]:
casos_16 = data[
    (data['UnitPrice'] == 0) &
    (data['Description'].isnull()) &
    (data['Quantity'] > 0) &
    (~data['StockCode'].isin(
        data[
            (data['UnitPrice'] > 0) &
            (data['Description'].notnull()) &
            (data['Quantity'] > 0)
        ]['StockCode']
    ))
]

### Revisión de los 16 registros sin descripción identificable

Se revisan los 16 registros cuyo precio por unidad es cero, cuya descripción está ausente y cuyo código de producto no pudo asociarse con una descripción disponible en transacciones con precio mayor que cero.

Estos registros presentan características atípicas y se mantienen como casos especiales para definir su tratamiento durante la limpieza.

In [26]:
data[data['InvoiceNo'].isin(casos_16['InvoiceNo'])]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,NaN,United Kingdom
7187,536995,35951,NaN,57,2010-12-03 15:29:00,0.0,NaN,United Kingdom
19628,537875,20849,NaN,1,2010-12-08 18:08:00,0.0,NaN,United Kingdom
19631,537878,72803B,NaN,1,2010-12-08 18:09:00,0.0,NaN,United Kingdom
21782,538133,85018C,NaN,3,2010-12-09 15:56:00,0.0,NaN,United Kingdom
21786,538137,62095B,NaN,2,2010-12-09 15:57:00,0.0,NaN,United Kingdom


### Identificación de registros con información de cliente disponible

Se identifican los registros con precio por unidad igual a cero que cuentan tanto con una descripción del producto como con un identificador de cliente.

Este subconjunto permite diferenciar los casos que, a pesar de tener precio cero, conservan información suficiente sobre el producto y el cliente para realizar una revisión específica.

In [27]:
# Casos donde el precio es 0, pero la descripción y el CustomerID no son nulos 
caso_40 = data[(data['UnitPrice'] == 0) & (~data['Description'].isnull()) &(~data['CustomerID'].isnull())]

### Identificación de los registros restantes

Se identifican los registros con precio por unidad igual a cero y descripción disponible que no pertenecen al subconjunto de registros que cuentan con un `CustomerID`.

Estos registros conforman el grupo restante de casos que será analizado para determinar su naturaleza y definir su tratamiento durante la limpieza de los datos.

In [28]:
#
caso_1021 = data[
    (data['UnitPrice'] == 0) &
    (~data['Description'].isnull()) &
    (data['CustomerID'].isnull())
]

### Cantidad de registros identificados

Se verifica la cantidad de registros que cumplen con las condiciones establecidas para `caso_1021`.

Esta comprobación permite confirmar el número de observaciones que serán analizadas posteriormente.

In [29]:
caso_1021.shape

(1021, 8)




## Análisis del Caso 40

A continuación se analiza de manera específica el subconjunto de 40 registros identificado anteriormente.

Estos registros presentan un precio por unidad igual a cero, pero cuentan tanto con una descripción del producto como con un identificador de cliente. Por esta razón, se revisan por separado para determinar si corresponden a operaciones válidas o a registros que requieren un tratamiento especial durante la limpieza.


In [30]:
caso_40

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
9302,537197,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 14:02:00,0.0,12647.0,Germany
33576,539263,22580,ADVENT CALENDAR GINGHAM SACK,4,2010-12-16 14:36:00,0.0,16560.0,United Kingdom
40089,539722,22423,REGENCY CAKESTAND 3 TIER,10,2010-12-21 13:45:00,0.0,14911.0,EIRE
47068,540372,22090,PAPER BUNTING RETROSPOT,24,2011-01-06 16:41:00,0.0,13081.0,United Kingdom
47070,540372,22553,PLASTERS IN TIN SKULLS,24,2011-01-06 16:41:00,0.0,13081.0,United Kingdom
56674,541109,22168,ORGANISER WOOD ANTIQUE WHITE,1,2011-01-13 15:10:00,0.0,15107.0,United Kingdom
86789,543599,84535B,FAIRY CAKES NOTEBOOK A6 SIZE,16,2011-02-10 13:08:00,0.0,17560.0,United Kingdom
130188,547417,22062,CERAMIC BOWL WITH LOVE HEART DESIGN,36,2011-03-23 10:25:00,0.0,13239.0,United Kingdom
139453,548318,22055,MINI CAKE STAND HANGING STRAWBERY,5,2011-03-30 12:45:00,0.0,13113.0,United Kingdom
145208,548871,22162,HEART GARLAND RUSTIC PADDED,2,2011-04-04 14:42:00,0.0,14410.0,United Kingdom


### Verificación de productos con precio válido

Se comprueba si el código de producto `22841`, identificado dentro del Caso 40, también aparece en otras transacciones con un precio por unidad mayor que cero.

El resultado muestra que este producto cuenta con registros con precios disponibles, lo que permite comparar estos casos con las transacciones en las que presenta un precio igual a cero.

In [31]:
data [
(data['StockCode']== 22841) &
(data['UnitPrice'] > 0)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2834,536592,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-01 17:06:00,16.13,NaN,United Kingdom
3925,536738,22841,ROUND CAKE TIN VINTAGE GREEN,2,2010-12-02 13:06:00,7.95,15827.0,United Kingdom
5111,536844,22841,ROUND CAKE TIN VINTAGE GREEN,2,2010-12-02 18:49:00,7.95,14236.0,United Kingdom
5328,536859,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-03 10:43:00,7.95,12841.0,United Kingdom
9543,537207,22841,ROUND CAKE TIN VINTAGE GREEN,1,2010-12-05 15:06:00,7.95,17402.0,United Kingdom
...,...,...,...,...,...,...,...,...
516920,580021,22841,ROUND CAKE TIN VINTAGE GREEN,2,2011-12-01 11:43:00,7.95,14047.0,United Kingdom
528422,580729,22841,ROUND CAKE TIN VINTAGE GREEN,1,2011-12-05 17:24:00,15.79,NaN,United Kingdom
535728,581217,22841,ROUND CAKE TIN VINTAGE GREEN,1,2011-12-08 09:20:00,15.79,NaN,United Kingdom
536171,581219,22841,ROUND CAKE TIN VINTAGE GREEN,1,2011-12-08 09:28:00,15.79,NaN,United Kingdom


### Cantidad de productos diferentes en el Caso 40

Se determina la cantidad de códigos de producto (`StockCode`) diferentes presentes en los registros del Caso 40.

Esta medida permite conocer si los 40 registros corresponden a pocos productos repetidos o si involucran una mayor variedad de productos.

In [32]:
caso_40['StockCode'].nunique()

35

### Identificación de los códigos de producto

Se identifican los códigos de producto presentes en el Caso 40 y se cuenta cuántos registros corresponden a cada uno.

In [33]:
caso_40['StockCode'].value_counts()

StockCode
M         6
22841     1
23268     1
21786     1
PADS      1
22778     1
21208     1
23234     1
22065     1
22385     1
22625     1
22899     1
23480     1
22437     1
22089     1
22464     1
23407     1
22955     1
23270     1
22580     1
23157     1
22960     1
22167     1
22619     1
47566     1
22636     1
22162     1
22055     1
22062     1
84535B    1
22168     1
22553     1
22090     1
22423     1
84826     1
Name: count, dtype: int64

In [34]:
# Comprobamos si M aparece con UnitPrice > 0
data [
(data['StockCode'] == 'M') 
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2239,536569,M,Manual,1,2010-12-01 15:35:00,1.25,16274.0,United Kingdom
2250,536569,M,Manual,1,2010-12-01 15:35:00,18.95,16274.0,United Kingdom
5684,536865,M,Manual,1,2010-12-03 11:28:00,2.55,NaN,United Kingdom
6798,536981,M,Manual,2,2010-12-03 14:26:00,0.85,14723.0,United Kingdom
7976,537077,M,Manual,12,2010-12-05 11:59:00,0.42,17062.0,United Kingdom
...,...,...,...,...,...,...,...,...
531432,580956,M,Manual,4,2011-12-06 14:23:00,1.25,17841.0,United Kingdom
533082,C581009,M,Manual,-1,2011-12-07 09:15:00,125.00,16971.0,United Kingdom
534308,C581145,M,Manual,-1,2011-12-07 13:48:00,9.95,17490.0,United Kingdom
538321,581405,M,Manual,3,2011-12-08 13:50:00,0.42,13521.0,United Kingdom


In [35]:
# Comprobamos si todos los registros con StockCode M tienen descripción Manual
data [
(data['StockCode'] == 'M') &
(data['Description'] == 'Manual')
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2239,536569,M,Manual,1,2010-12-01 15:35:00,1.25,16274.0,United Kingdom
2250,536569,M,Manual,1,2010-12-01 15:35:00,18.95,16274.0,United Kingdom
5684,536865,M,Manual,1,2010-12-03 11:28:00,2.55,NaN,United Kingdom
6798,536981,M,Manual,2,2010-12-03 14:26:00,0.85,14723.0,United Kingdom
7976,537077,M,Manual,12,2010-12-05 11:59:00,0.42,17062.0,United Kingdom
...,...,...,...,...,...,...,...,...
531432,580956,M,Manual,4,2011-12-06 14:23:00,1.25,17841.0,United Kingdom
533082,C581009,M,Manual,-1,2011-12-07 09:15:00,125.00,16971.0,United Kingdom
534308,C581145,M,Manual,-1,2011-12-07 13:48:00,9.95,17490.0,United Kingdom
538321,581405,M,Manual,3,2011-12-08 13:50:00,0.42,13521.0,United Kingdom


In [36]:
# Verificamos caso_40 STOCKCODE
caso_40[caso_40['StockCode'] == 'M']

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
285657,561916,M,Manual,1,2011-08-01 11:44:00,0.0,15581.0,United Kingdom
361825,568384,M,Manual,1,2011-09-27 09:46:00,0.0,12748.0,United Kingdom
395529,571035,M,Manual,1,2011-10-13 12:50:00,0.0,12446.0,RSA
436961,574252,M,Manual,1,2011-11-03 13:24:00,0.0,12437.0,France
479546,577168,M,Manual,1,2011-11-18 10:42:00,0.0,12603.0,Germany
485985,577696,M,Manual,1,2011-11-21 11:57:00,0.0,16406.0,United Kingdom


In [37]:
# Comprobamos la frecuencia de cada InvoiceNo en caso_40
caso_40['InvoiceNo'].value_counts()

InvoiceNo
564651    4
540372    2
575579    2
574920    2
537197    1
574252    1
571035    1
572893    1
574138    1
574175    1
574879    1
574469    1
568384    1
577129    1
577168    1
577314    1
577696    1
569716    1
568158    1
539263    1
562973    1
561916    1
561669    1
561284    1
554037    1
553000    1
550188    1
548871    1
548318    1
547417    1
543599    1
541109    1
539722    1
578841    1
Name: count, dtype: int64

In [38]:
# Comprobamos si existen otros registros con el mismo InvoiceNo en data
data[
(data['InvoiceNo'] == 564651)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
314745,564651,23270,SET OF 2 CERAMIC PAINTED HEARTS,96,2011-08-26 14:19:00,0.0,14646.0,Netherlands
314746,564651,23268,SET OF 2 CERAMIC CHRISTMAS REINDEER,192,2011-08-26 14:19:00,0.0,14646.0,Netherlands
314747,564651,22955,36 FOIL STAR CAKE CASES,144,2011-08-26 14:19:00,0.0,14646.0,Netherlands
314748,564651,21786,POLKADOT RAIN HAT,144,2011-08-26 14:19:00,0.0,14646.0,Netherlands


In [39]:
# Comprobamos otro InvoiceNo en data
data[
(data['InvoiceNo'] == 540372)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
46900,540372,21726,MULTI HEARTS STICKERS,12,2011-01-06 16:41:00,0.85,13081.0,United Kingdom
46901,540372,21723,ALPHABET HEARTS STICKER SHEET,12,2011-01-06 16:41:00,0.85,13081.0,United Kingdom
46902,540372,22934,BAKING MOULD EASTER EGG WHITE CHOC,6,2011-01-06 16:41:00,2.95,13081.0,United Kingdom
46903,540372,22935,BAKING MOULD ROSE MILK CHOCOLATE,6,2011-01-06 16:41:00,3.25,13081.0,United Kingdom
46904,540372,22936,BAKING MOULD ROSE WHITE CHOCOLATE,6,2011-01-06 16:41:00,3.25,13081.0,United Kingdom
...,...,...,...,...,...,...,...,...
47066,540372,22961,JAM MAKING SET PRINTED,12,2011-01-06 16:41:00,1.45,13081.0,United Kingdom
47067,540372,22960,JAM MAKING SET WITH JARS,6,2011-01-06 16:41:00,4.25,13081.0,United Kingdom
47068,540372,22090,PAPER BUNTING RETROSPOT,24,2011-01-06 16:41:00,0.00,13081.0,United Kingdom
47069,540372,22432,WATERING CAN PINK BUNNY,18,2011-01-06 16:41:00,1.95,13081.0,United Kingdom


### Verificación de precios de los productos del Caso 40

Se comprueba si los productos identificados en el Caso 40 también aparecen en el resto del conjunto de datos con un precio por unidad mayor que cero.

In [40]:
data[
    data['StockCode'].isin(caso_40['StockCode']) &
    (data['UnitPrice'] > 0)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
21,536368,22960,JAM MAKING SET WITH JARS,6,2010-12-01 08:34:00,4.25,13047.0,United Kingdom
138,536381,22778,GLASS CLOCHE SMALL,3,2010-12-01 09:41:00,3.95,15311.0,United Kingdom
157,536384,22464,HANGING METAL HEART LANTERN,12,2010-12-01 09:53:00,1.65,18074.0,United Kingdom
170,536385,22960,JAM MAKING SET WITH JARS,6,2010-12-01 09:56:00,4.25,17420.0,United Kingdom
173,536385,22168,ORGANISER WOOD ANTIQUE WHITE,2,2010-12-01 09:56:00,8.50,17420.0,United Kingdom
...,...,...,...,...,...,...,...,...
541685,581538,21208,PASTEL COLOUR HONEYCOMB FAN,1,2011-12-09 11:34:00,0.39,14446.0,United Kingdom
541704,581567,22464,HANGING METAL HEART LANTERN,24,2011-12-09 11:56:00,0.79,16626.0,United Kingdom
541709,581567,22625,RED KITCHEN SCALES,2,2011-12-09 11:56:00,8.50,16626.0,United Kingdom
541823,581579,23480,MINI LIGHTS WOODLAND MUSHROOMS,8,2011-12-09 12:19:00,3.75,17581.0,United Kingdom


### Verificación de los clientes del Caso 40

Se comprueba si los clientes asociados al Caso 40 también aparecen en otras transacciones del conjunto de datos.

In [41]:
data[
    data['CustomerID'].isin(caso_40['CustomerID'])
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
197,536389,22941,CHRISTMAS LIGHTS 10 REINDEER,6,2010-12-01 10:03:00,8.50,12431.0,Australia
198,536389,21622,VINTAGE UNION JACK CUSHION COVER,8,2010-12-01 10:03:00,4.95,12431.0,Australia
199,536389,21791,VINTAGE HEADS AND TAILS CARD GAME,12,2010-12-01 10:03:00,1.25,12431.0,Australia
200,536389,35004C,SET OF 3 COLOURED FLYING DUCKS,6,2010-12-01 10:03:00,5.45,12431.0,Australia
201,536389,35004G,SET OF 3 GOLD FLYING DUCKS,4,2010-12-01 10:03:00,6.35,12431.0,Australia
...,...,...,...,...,...,...,...,...
541889,581585,22466,FAIRY TALE COTTAGE NIGHT LIGHT,12,2011-12-09 12:31:00,1.95,15804.0,United Kingdom
541890,581586,22061,LARGE CAKE STAND HANGING STRAWBERY,8,2011-12-09 12:49:00,2.95,13113.0,United Kingdom
541891,581586,23275,SET OF 3 HANGING OWLS OLLIE BEAK,24,2011-12-09 12:49:00,1.25,13113.0,United Kingdom
541892,581586,21217,RED RETROSPOT ROUND CAKE TINS,24,2011-12-09 12:49:00,8.95,13113.0,United Kingdom


### Revisión de otras facturas

In [42]:
# Comprobamos otro InvoiceNo para identificar posibles patrones
data[
(data['InvoiceNo'] == 575579)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
454463,575579,22437,SET OF 9 BLACK SKULL BALLOONS,20,2011-11-10 11:49:00,0.00,13081.0,United Kingdom
454464,575579,22089,PAPER BUNTING VINTAGE PAISLEY,24,2011-11-10 11:49:00,0.00,13081.0,United Kingdom
454465,575579,22197,POPCORN HOLDER,100,2011-11-10 11:49:00,0.72,13081.0,United Kingdom


In [43]:
# Comprobamos otro InvoiceNo para identificar posibles patrones
data[
(data['InvoiceNo'] == 574920)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
446773,574920,85099B,JUMBO BAG RED RETROSPOT,10,2011-11-07 16:34:00,2.08,13985.0,United Kingdom
446774,574920,22663,JUMBO BAG DOLLY GIRL DESIGN,10,2011-11-07 16:34:00,2.08,13985.0,United Kingdom
446775,574920,22385,JUMBO BAG SPACEBOY DESIGN,10,2011-11-07 16:34:00,2.08,13985.0,United Kingdom
446776,574920,22386,JUMBO BAG PINK POLKADOT,10,2011-11-07 16:34:00,2.08,13985.0,United Kingdom
446777,574920,22457,NATURAL SLATE HEART CHALKBOARD,6,2011-11-07 16:34:00,2.95,13985.0,United Kingdom
446778,574920,22083,PAPER CHAIN KIT RETROSPOT,6,2011-11-07 16:34:00,2.95,13985.0,United Kingdom
446779,574920,22910,PAPER CHAIN KIT VINTAGE CHRISTMAS,12,2011-11-07 16:34:00,2.95,13985.0,United Kingdom
446780,574920,23313,VINTAGE CHRISTMAS BUNTING,10,2011-11-07 16:34:00,4.95,13985.0,United Kingdom
446781,574920,23480,MINI LIGHTS WOODLAND MUSHROOMS,4,2011-11-07 16:34:00,3.75,13985.0,United Kingdom
446782,574920,20727,LUNCH BAG BLACK SKULL.,10,2011-11-07 16:34:00,1.65,13985.0,United Kingdom


In [44]:
# Comprobamos otro InvoiceNo para identificar posibles patrones
data[
(data['InvoiceNo'] == 537197)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
9287,537197,22843,BISCUIT TIN VINTAGE GREEN,2,2010-12-05 14:02:00,6.75,12647.0,Germany
9288,537197,22842,BISCUIT TIN VINTAGE RED,2,2010-12-05 14:02:00,6.75,12647.0,Germany
9289,537197,22961,JAM MAKING SET PRINTED,12,2010-12-05 14:02:00,1.45,12647.0,Germany
9290,537197,22962,JAM JAR WITH PINK LID,12,2010-12-05 14:02:00,0.85,12647.0,Germany
9291,537197,22963,JAM JAR WITH GREEN LID,12,2010-12-05 14:02:00,0.85,12647.0,Germany
9292,537197,84997D,PINK 3 PIECE POLKADOT CUTLERY SET,6,2010-12-05 14:02:00,3.75,12647.0,Germany
9293,537197,22199,FRYING PAN RED RETROSPOT,4,2010-12-05 14:02:00,4.25,12647.0,Germany
9294,537197,22200,FRYING PAN PINK POLKADOT,4,2010-12-05 14:02:00,4.25,12647.0,Germany
9295,537197,22505,MEMO BOARD COTTAGE DESIGN,4,2010-12-05 14:02:00,4.95,12647.0,Germany
9296,537197,22507,MEMO BOARD RETROSPOT DESIGN,4,2010-12-05 14:02:00,4.95,12647.0,Germany


### Conclusión del Caso 40

Los 40 registros con UnitPrice = 0, Description presente y CustomerID presente presentan características compatibles con transacciones reales. Las facturas asociadas contienen tanto líneas con precios normales como líneas con precio 0, y los mismos productos aparecen en otras transacciones con precios distintos de cero.

No se encontró evidencia suficiente para considerarlos registros erróneos, por lo que se conservan sin modificar en esta etapa.






## REVISAMOS EL CASO 1021 
Correspondiente a : Descripción presente, precio 0 y Cliente nulo 





In [45]:
caso_1021

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
6391,536941,22734,amazon,20,2010-12-03 12:08:00,0.0,NaN,United Kingdom
6392,536942,22139,amazon,15,2010-12-03 12:08:00,0.0,NaN,United Kingdom
7313,537032,21275,?,-30,2010-12-03 16:50:00,0.0,NaN,United Kingdom
13217,537425,84968F,check,-20,2010-12-06 15:35:00,0.0,NaN,United Kingdom
13218,537426,84968E,check,-35,2010-12-06 15:36:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
535336,581213,22576,check,-30,2011-12-07 18:38:00,0.0,NaN,United Kingdom
536908,581226,23090,missing,-338,2011-12-08 09:56:00,0.0,NaN,United Kingdom
538504,581406,46000M,POLYESTER FILLER PAD 45x45cm,240,2011-12-08 13:58:00,0.0,NaN,United Kingdom
538505,581406,46000S,POLYESTER FILLER PAD 40x40cm,300,2011-12-08 13:58:00,0.0,NaN,United Kingdom


In [46]:
# Analizamos las descripciones y su frecuencia en el caso_1021
caso_1021['Description'].value_counts()

Description
check                               159
?                                    47
damages                              45
damaged                              43
found                                25
                                   ... 
cracked                               1
sold as 22467                         1
CHICK GREY HOT WATER BOTTLE           1
mystery! Only ever imported 1800      1
lost                                  1
Name: count, Length: 345, dtype: int64

In [47]:
# Revisamos un tipo de descripción específico dentro del caso_1021
caso_1021[caso_1021['Description'] == 'CHICK GREY HOT WATER BOTTLE']

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
147243,549100,21484,CHICK GREY HOT WATER BOTTLE,1,2011-04-06 11:14:00,0.0,NaN,United Kingdom


In [48]:
# Buscamos descripciones relacionadas con HOT WATER BOTTLE

caso_1021[caso_1021['Description'].str.contains(
    'HOT WATER BOTTLE',
    case=False,
    na=False
)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
30603,538877,21479,WHITE SKULL HOT WATER BOTTLE,1,2010-12-14 15:29:00,0.0,NaN,United Kingdom
30644,538877,22112,CHOCOLATE HOT WATER BOTTLE,1,2010-12-14 15:29:00,0.0,NaN,United Kingdom
40952,539750,22114,HOT WATER BOTTLE TEA AND SYMPATHY,1,2010-12-21 15:40:00,0.0,NaN,United Kingdom
41094,539750,21481,FAWN BLUE HOT WATER BOTTLE,1,2010-12-21 15:40:00,0.0,NaN,United Kingdom
147243,549100,21484,CHICK GREY HOT WATER BOTTLE,1,2011-04-06 11:14:00,0.0,NaN,United Kingdom


In [49]:
# Verificamos si el producto tiene registros con precio > 0 en el dataset
data[
    (data['StockCode'] == 21479) &
    (data['UnitPrice'] > 0)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
313,536398,21479,WHITE SKULL HOT WATER BOTTLE,4,2010-12-01 10:52:00,3.75,13448.0,United Kingdom
445,536408,21479,WHITE SKULL HOT WATER BOTTLE,4,2010-12-01 11:41:00,3.75,14307.0,United Kingdom
484,536409,21479,WHITE SKULL HOT WATER BOTTLE,1,2010-12-01 11:45:00,3.75,17908.0,United Kingdom
706,536423,21479,WHITE SKULL HOT WATER BOTTLE,8,2010-12-01 12:08:00,3.75,18085.0,United Kingdom
1946,536544,21479,WHITE SKULL HOT WATER BOTTLE,5,2010-12-01 14:32:00,7.62,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
539446,581439,21479,WHITE SKULL HOT WATER BOTTLE,1,2011-12-08 16:30:00,8.29,NaN,United Kingdom
540330,581476,21479,WHITE SKULL HOT WATER BOTTLE,4,2011-12-09 08:48:00,4.25,12433.0,Norway
540630,581492,21479,WHITE SKULL HOT WATER BOTTLE,2,2011-12-09 10:03:00,8.29,NaN,United Kingdom
541271,581497,21479,WHITE SKULL HOT WATER BOTTLE,25,2011-12-09 10:23:00,9.13,NaN,United Kingdom


In [50]:
# Verificamos si el producto tiene registros con precio > 0 en el dataset
data[
    (data['StockCode'] == 22112) &
    (data['UnitPrice'] > 0)
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
312,536398,22112,CHOCOLATE HOT WATER BOTTLE,9,2010-12-01 10:52:00,4.95,13448.0,United Kingdom
540,536409,22112,CHOCOLATE HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
1319,536534,22112,CHOCOLATE HOT WATER BOTTLE,3,2010-12-01 13:33:00,4.95,15350.0,United Kingdom
1510,536544,22112,CHOCOLATE HOT WATER BOTTLE,2,2010-12-01 14:32:00,11.02,NaN,United Kingdom
1998,536551,22112,CHOCOLATE HOT WATER BOTTLE,1,2010-12-01 14:34:00,4.95,17346.0,United Kingdom
...,...,...,...,...,...,...,...,...
540397,581480,22112,CHOCOLATE HOT WATER BOTTLE,6,2011-12-09 09:07:00,4.95,14441.0,United Kingdom
540739,581492,22112,CHOCOLATE HOT WATER BOTTLE,1,2011-12-09 10:03:00,10.79,NaN,United Kingdom
541246,581496,22112,CHOCOLATE HOT WATER BOTTLE,6,2011-12-09 10:20:00,4.95,16558.0,United Kingdom
541283,581497,22112,CHOCOLATE HOT WATER BOTTLE,13,2011-12-09 10:23:00,12.46,NaN,United Kingdom


In [51]:
# Comprobamos qué proporción de los 1021 registros tiene un StockCode que también aparece con precio > 0
caso_1021['StockCode'].isin(
    data[data['UnitPrice'] > 0]['StockCode']
).value_counts()

StockCode
True     1000
False      21
Name: count, dtype: int64

De los 1,021 registros, 1,000 tienen un StockCode que también aparece en otras transacciones con UnitPrice > 0, mientras que 21 no presentan esta coincidencia.

In [52]:
# Revisamos los 21 productos que no aparecen con precio > 0
caso_1021[
    ~caso_1021['StockCode'].isin(
        data[data['UnitPrice'] > 0]['StockCode']
    )
]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
51049,540638,72038P,damages,-990,2011-01-10 12:14:00,0.0,NaN,United Kingdom
82794,543257,84611B,thrown away,-1430,2011-02-04 16:06:00,0.0,NaN,United Kingdom
82796,543259,84612B,thrown away,-162,2011-02-04 16:07:00,0.0,NaN,United Kingdom
115489,546126,35611B,thrown away,-27,2011-03-09 14:52:00,0.0,NaN,United Kingdom
115818,546152,72140F,throw away,-5368,2011-03-09 17:25:00,0.0,NaN,United Kingdom
131398,547559,72759,thrown away-can't sell.,-524,2011-03-23 17:27:00,0.0,NaN,United Kingdom
131399,547560,72732,thrown away-can't sell,-2472,2011-03-23 17:28:00,0.0,NaN,United Kingdom
185113,552733,23059,Thrown away-rusty,-2376,2011-05-11 10:48:00,0.0,NaN,United Kingdom
254130,559296,23001,TRAVEL CARD WALLET DOTCOMGIFTSHOP,200,2011-07-07 12:35:00,0.0,NaN,United Kingdom
279251,561249,DCGS0073,ebay,-4,2011-07-26 11:51:00,0.0,NaN,United Kingdom


In [53]:
# Verificamos las cantidades registradas en el caso_1021
caso_1021['Quantity']

6391       20
6392       15
7313      -30
13217     -20
13218     -35
         ... 
535336    -30
536908   -338
538504    240
538505    300
538919   -235
Name: Quantity, Length: 1021, dtype: int64

In [54]:
# Analizamos las descripciones asociadas a cantidades positivas
caso_1021[caso_1021['Quantity'] > 0]['Description'].value_counts()

Description
check                                 39
found                                 25
adjustment                            14
FRENCH BLUE METAL DOOR SIGN 1          9
amazon                                 8
                                      ..
CREAM SWEETHEART LETTER RACK           1
VICTORIAN SEWING BOX LARGE             1
RED RETROSPOT BUTTER DISH              1
VINTAGE HEADS AND TAILS CARD GAME      1
CLASSIC GLASS COOKIE JAR               1
Name: count, Length: 215, dtype: int64

In [55]:
## Analizamos las descripciones asociadas a los registros con cantidades negativas
caso_1021[caso_1021['Quantity'] < 0]['Description'].value_counts()

Description
check                                  120
damages                                 45
damaged                                 42
?                                       41
sold as set on dotcom                   20
                                      ... 
POSSIBLE DAMAGES OR LOST?                1
MERCHANT CHANDLER CREDIT ERROR, STO      1
mystery! Only ever imported 1800         1
sold as 22467                            1
lost                                     1
Name: count, Length: 138, dtype: int64

In [56]:
# Agrupamos por descripción y analizamos frecuencia y cantidad total
caso_1021.groupby('Description')['Quantity'].agg(['count', 'sum']).sort_values('count', ascending=False)

,count,sum
Description,,
check,159,-12030
?,47,-1384
damages,45,-4338
damaged,43,-1632
found,25,1990
...,...,...
OOPS ! adjustment,1,-384
PACK OF 12 LONDON TISSUES,1,204
PACK OF 6 BIRDY GIFT TAGS,1,1


### Conclusión del Caso 1021

Los 1,021 registros identificados presentan `UnitPrice = 0`, `CustomerID` ausente y `Description` disponible. La revisión mostró que la mayoría de sus `StockCode` también aparecen en otras transacciones con precio mayor que cero, mientras que una parte menor no pudo asociarse con operaciones comerciales normales.

En conjunto, estos registros presentan características diferentes de una venta comercial convencional y se consideran parte de un grupo de operaciones que requiere exclusión del análisis de ventas.

### Revisión de registros con precio cero y descripción faltante

Después de analizar los registros con `UnitPrice = 0`, se revisó el grupo que presenta simultáneamente `Description` y `CustomerID` ausentes.

Se identificaron **1,454 registros** con estas características. Al considerar también los **1,021 registros** analizados previamente en el Caso 1021, se observa que existen **2,475 registros** con `UnitPrice = 0` y `CustomerID` ausente.

Esta revisión permite determinar si estos registros corresponden a operaciones comerciales o si presentan características que justifican su exclusión del dataset procesado.

In [57]:
# Calculamos las unidades asociadas a los registros con precio cero,
# descripción faltante y cantidad positiva
data[
    (data['UnitPrice'] == 0) &
    (data['Description'].isnull()) &
    (data['Quantity'] > 0)
]['Quantity'].sum()

32547

In [58]:
# Comprobamos todos los registros con precio cero y cliente sin identificar
(
    (data['UnitPrice'] == 0) &
    (data['CustomerID'].isnull())
).sum()
# Revisamos cuántos tienen descripción y cuántos no
data[
    (data['UnitPrice'] == 0) &
    (data['CustomerID'].isnull())
]['Description'].isnull().value_counts()

Description
True     1454
False    1021
Name: count, dtype: int64

In [59]:
# Calculamos las unidades positivas asociadas a los registros
# con precio cero y descripción faltante, excluyendo códigos especiales
data[
    (data['UnitPrice'] == 0) &
    (data['Description'].isnull()) &
    (data['Quantity'] > 0) &
    (~data['StockCode'].isin(['DOT', 'POST', 'M', 'm', 'S', 'PADS'])) &
    (~data['StockCode'].str.startswith('gift_', na=False))
]['Quantity'].sum()

28137

### Impacto en el análisis de unidades

Aunque estos registros no generan ingresos debido a que `UnitPrice = 0`, algunos contienen cantidades positivas. Después de excluir los códigos especiales identificados, se contabilizaron **28,137 unidades** asociadas a registros sin descripción.

Estas unidades podrían afectar análisis basados en la cantidad vendida, como la rotación de productos y la evolución de la demanda. Por este motivo, se considera más consistente excluir del dataset procesado los registros que presentan simultáneamente `UnitPrice = 0` y `CustomerID` ausente.

### Conclusión del proceso de revisión

La exploración permitió identificar un total de **2,475 registros con `UnitPrice = 0` y `CustomerID` ausente**. Este grupo incluye tanto los 1,021 registros analizados inicialmente como los 1,454 registros sin descripción.

Debido a que estos registros no presentan un precio de venta ni un identificador de cliente y pueden introducir cantidades que no corresponden a ventas comerciales convencionales, se decide excluirlos del dataset procesado.

Esta decisión se aplicará de forma general en el script `clean_data.py`, en lugar de tratar los 1,021 registros como un caso aislado.